### PR TITLE
Pc/hpo

### PR DESCRIPTION
**Pull request by Blaizzy**
# Description

This PR adds a basic HPO example.

## ❔ This change

- [x] adds a new feature
- [ ] fixes breaking code
- [ ] is cosmetic (refactoring/reformatting)

---

## ✔️ Pre-merge checklist

Mark which items have been completed:

- [x] Refactored code ([sourcery](https://sourcery.ai/))
- [x] Tested code locally
- [x] Added code to GitHub tests ([notebooks](workflows/test-notebooks.yml), [scripts](workflows/test-scripts.yml))
- [x] Updated GitHub [README](../README.md)

### 🪝 Run the following to both [pre-commit](https://pre-commit.com/) hooks and JupyterLab (turn on `format on save`)

- [x] Stripped outputs from notebooks - ([nbstripout](https://pypi.org/project/nbstripout/))
- [x] Sorted imports - ([isort](https://pycqa.github.io/isort/))
- [x] Formatted code - ([back](https://github.com/psf/black))

### 📄 Update documentation

Coordinate with @normandy7

- [ ] Update content in the [docs](https://docs.neptune.ai)
- [ ] (For new content) Add links to examples [README](../README.md)

---

## 🧪 Test Configuration

- OS:
- Python version:
- Neptune version:
- Affected libraries with version:

---

## ✔️ Post-merge checklist

These tasks should be done after the code is merged to main. Create a subtask under the main ClickUp task to have these on your radar and share the subtask here.

__Post-merge subtask:__ `(ClickUp/JIRA task name)`

- [x] Update relevant blogs with impacted code/functionality (Contact @Patrycja-J for the list of blogs)
- [ ] Add functionality under the Neptune column of the `competitor comparison sheet` (Coordinate with @SiddhantSadangi)
- [ ] Update integration README (Contact @Patrycja-J)

### CODE REVIEW

File: how-to-guides/neptune-hpo/scripts/Neptune_hpo_separate_runs.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+
+# Log metadata from each trial into separate run
+for (i, lr) in enumerate(learning_rates):
+    # (Neptune) Create a run
+    run = neptune.init_run(
+        api_token=neptune.ANONYMOUS_API_TOKEN,
+        project="common/pytorch-integration",
+        name=f"trial-{i}",
+    )
```
Comment (SiddhantSadangi) [inline how-to-guides/neptune-hpo/scripts/Neptune_hpo_separate_runs.py:75]:
```suggestion
        name=f"trial-{i}",
        tags=["trial-level"],
    )
```

File: how-to-guides/neptune-hpo/scripts/Neptune_hpo_single_run.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+import torch
+import torch.nn as nn
+import torch.optim as optim
+from torchvision import datasets, transforms
+
+# (Neptune) Create a run
+run = neptune.init_run(
+    api_token=neptune.ANONYMOUS_API_TOKEN,
+    project="common/pytorch-integration",
+)
```
Comment (SiddhantSadangi) [inline how-to-guides/neptune-hpo/scripts/Neptune_hpo_single_run.py:11]:
```suggestion
    project="common/pytorch-integration",
    tags=["sweep-level"],
)
```

File: how-to-guides/neptune-hpo/notebooks/Neptune_hpo.ipynb [hunk 3]
```diff
@@ -0,0 +1,386 @@
...
+    "# How to use Neptune in HPO training job"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Introduction\n",
+    "\n",
+    "When running a hyperparameter optimization job, you can use neptune to track all the metadata from the study and each trial.\n",
```
Comment (normandy7) [inline how-to-guides/neptune-hpo/notebooks/Neptune_hpo.ipynb:16]:
```suggestion
    "When running a hyperparameter optimization job, you can use Neptune to track all the metadata from the study and each trial.\n",
```

File: how-to-guides/neptune-hpo/notebooks/Neptune_hpo.ipynb [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+    "\n",
+    "By the end of this guide, you will learn how to configure neptune to track your hyperparameter optimization job’s metadata.\n",
```
Comment (normandy7) [inline how-to-guides/neptune-hpo/notebooks/Neptune_hpo.ipynb:18]:
```suggestion
    "In this guide, you'll learn how to configure Neptune to track the metadata of your hyperparameter optimization job.\n",
```

File: how-to-guides/neptune-hpo/notebooks/Neptune_hpo.ipynb [hunk 5]
```diff
@@ -0,0 +1,386 @@
...
+    "import torch.optim as optim\n",
+    "from torchvision import datasets, transforms"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Log metadata across HPO trials into a single run\n",
+    "Create a global neptune run to log metadata (i.e. metrics) across different trials.\n",
```
Comment (normandy7) [inline how-to-guides/neptune-hpo/notebooks/Neptune_hpo.ipynb:79]:
```suggestion
    "Create a global Neptune run to log metadata (i.e. metrics) across different trials.\n",
```

File: how-to-guides/neptune-hpo/notebooks/Neptune_hpo.ipynb [hunk 6]
```diff
@@ -0,0 +1,386 @@
...
+    "# Stop logging\n",
+    "run.stop()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Log metadata from each HPO trial into separate runs\n",
+    "Create a local neptune runs to log metrics from each trial into separate run"
```
Comment (normandy7) [inline how-to-guides/neptune-hpo/notebooks/Neptune_hpo.ipynb:302]: One or multiple? 

```suggestion
    "Create local Neptune runs to log metrics from each trial into separate run"
```

File: how-to-guides/neptune-hpo/notebooks/Neptune_hpo.ipynb [hunk 7]
```diff
@@ -0,0 +1,386 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Explore the results in Neptune\n",
+    "Follow the link to the run and explore metadata (such as metrics and hyperparameters) that were logged to the run in Neptune.\n",
+    "\n",
+    "You can also check out an example runs.\n",
```
Comment (normandy7) [inline how-to-guides/neptune-hpo/notebooks/Neptune_hpo.ipynb:359]:
```suggestion
    "You can also check out these example runs:\n",
```

**Review by normandy7 — CHANGES_REQUESTED** [review]

**Comment by Blaizzy** [comment]
@normandy7 could you please update the docs page link in the README?

**Comment by Blaizzy** [comment]
@SiddhantSadangi and @normandy7 all the changes are done.

**Comment by normandy7** [comment]
> @normandy7 could you please update the docs page link in the README?

Yes, I'll make sure to do this when the actual docs page is ready. You can disregard the docs checklist items from this PR.

**Comment by Blaizzy** [comment]
Got it. 👍

**Review by SiddhantSadangi — CHANGES_REQUESTED** [review]
I've made some changes to the present CV results with Neptune code as a part of #155 . 
Some of them can be adopted here too to reduce test time and prevent timeouts when running the scheduled jobs

**Comment by Blaizzy** [comment]
@SiddhantSadangi 

Could you please be a bit more specific?

**Comment by Blaizzy** [comment]
@SiddhantSadangi changes done 👍 

It now runs on FakeData

**Comment by SiddhantSadangi** [comment]
tqdm will need to be added to the requirements :)

**Comment by Blaizzy** [comment]
I have it locally by default 😆

**Comment by Blaizzy** [comment]
Done 👍

File: how-to-guides/neptune-hpo/notebooks/Neptune_hpo.ipynb [hunk 8]
```diff
@@ -0,0 +1,400 @@
...
+    "\n",
+    "            # Log metrics\n",
+    "            run[\"training/batch/loss\"].log(loss)\n",
+    "            run[\"training/batch/acc\"].log(acc)\n",
+    "\n",
+    "            loss.backward()\n",
+    "            optimizer.step()\n",
+    "\n",
+    "    # Wait for all the tracking calls to finish.\n",
+    "    run.wait()\n",
```
Comment (SiddhantSadangi) [inline how-to-guides/neptune-hpo/notebooks/Neptune_hpo.ipynb:361]: do we need to wait given that `run.stop()` will take care of syncing?

File: how-to-guides/neptune-hpo/scripts/Neptune_hpo_separate_runs.py [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+
+            # (Neptune) Log losses and metrics
+            run["training/batch/loss"].log(loss)
+            run["training/batch/acc"].log(acc)
+
+            loss.backward()
+            optimizer.step()
+
+    # (Neptune) Wait for all the tracking calls to finish.
+    run.wait()
```
Comment (SiddhantSadangi) [inline how-to-guides/neptune-hpo/scripts/Neptune_hpo_separate_runs.py:111]: do we need to wait given that `run.stop()` will take care of syncing?

**Review by SiddhantSadangi — CHANGES_REQUESTED** [review]

File: how-to-guides/neptune-hpo/notebooks/Neptune_hpo.ipynb (hunk 8 above)
Comment (Blaizzy) [reply how-to-guides/neptune-hpo/notebooks/Neptune_hpo.ipynb:361]: No, not really.

But I wanted to ensure all metadata gets logged before stopping.

File: how-to-guides/neptune-hpo/scripts/Neptune_hpo_separate_runs.py (hunk 9 above)
Comment (SiddhantSadangi) [reply how-to-guides/neptune-hpo/scripts/Neptune_hpo_separate_runs.py:111]: won't `run.stop()` do that before stopping the run?

File: how-to-guides/neptune-hpo/notebooks/Neptune_hpo.ipynb (hunk 8 above)
Comment (SiddhantSadangi) [reply how-to-guides/neptune-hpo/notebooks/Neptune_hpo.ipynb:361]: won't `run.stop()` do that before stopping the run?

**Review by SiddhantSadangi — APPROVED** [review]

File: how-to-guides/neptune-hpo/notebooks/Neptune_hpo.ipynb (hunk 8 above)
Comment (Blaizzy) [reply how-to-guides/neptune-hpo/notebooks/Neptune_hpo.ipynb]: Removed it.

**Review by normandy7 — APPROVED** [review]

**Review by SiddhantSadangi — APPROVED** [review]

**Pull request merged**